### PR TITLE
fix(cursor): route composer-2.5 tool continuations through userMessageAction

### PR DESCRIPTION
**Pull request by ZSN12**
## Summary

`cursor/composer-2.5` returns empty `content` with `finish_reason: stop` after tool-result continuation when proxied through opencodex. `cursor/composer-2.5-fast` works correctly on the same code path.

On `resumeAction`, composer-2.5 either issues native Cursor tool calls (read/grep/exec) instead of answering, or completes with zero visible text while consuming tokens.

**Fix:** Add `cursorNeedsExternalToolContinuation()` — routes `composer-2.5` tool-result continuations through `userMessageAction` + `CURSOR_EXTERNAL_TOOL_CONTINUATION_TEXT` instead of bare `resumeAction`. `composer-2.5-fast` unchanged.

**Files changed:**
- `src/adapters/cursor/discovery.ts` — new helper function
- `src/adapters/cursor/protobuf-request.ts` — use helper in action selection + history encoding

## Verification

Tested locally against opencodex proxy on port 10100:
- Before: 0/3 tool-continuation requests returned content
- After: 3/3 returned correct summary (streaming + non-streaming)
- Regression: `composer-2.5-fast`, plain chat, and first-turn tool calls unaffected

## Checklist

- [x] Scope stays focused and avoids unrelated cleanup.
- [x] Docs or release notes were updated when needed.
- [x] Security-sensitive changes were reviewed for secrets, auth, and unsafe defaults.

## Review readiness checklist

This PR stays in draft until every box below is ticked. Tick all four boxes once the requirements are met:

- [x] All CI tests are green on my local testing.
- [x] I pushed my PR to the latest dev commit.
- [x] I resolved all correct Codex and CodeRabbit findings.

- [x] My PR is ready for review.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved tool continuation handling for Cursor models.
  * Ensured supported Composer models use the correct continuation flow for tool results, conversations, diagnostics, and replayed requests.
  * Preserved native handling for Composer models that do not require external continuation.
  * Updated continuation behavior for Composer 2.5 while preserving existing behavior for its fast variant.

### CODE REVIEW

**Review by Ingwannu — CHANGES_REQUESTED** [review]
The reported `composer-2.5` continuation failure is worth fixing, but the current predicate is applied beyond the failing boundary and can change ordinary native conversation replay.

Two blockers on head `13ace17a3096a848a8812295c5ddeb4ed36481f0`:

1. `src/adapters/cursor/protobuf-request.ts` now uses `cursorNeedsExternalToolContinuation()` in `rootPromptMessages()` and `conversationTurns()`. Those functions run for ordinary turns as well as tool-result continuations. Because the predicate returns true for `composer-2.5`, normal native Composer requests can be replayed with the external-model history window and root-prompt behavior even when the last raw message is not a tool result. Keep `isCursorExternalWireModel()` for general history/replay semantics and apply the `composer-2.5` exception only when selecting the action for an actual trailing `toolResult`.
2. This changes runtime behavior under `src/` without regression coverage. Please add focused tests proving all three boundaries: `composer-2.5` plus a trailing tool result uses `userMessageAction`; `composer-2.5-fast` retains `resumeAction`; and an ordinary first/conversation turn for `composer-2.5` keeps the existing native replay semantics. The nearest suites are `tests/cursor-blob.test.ts`, `tests/cursor-tool-continuation.test.ts`, and `tests/cursor-request-builder.test.ts`.

I retargeted the PR from `main` to `dev`. Please rebase the branch onto the current `dev`, keep the exception narrow, add the tests above, rerun the focused Cursor suites plus the repository-required checks, and update the readiness checklist against the new exact head.

**Comment by lidge-jun** [comment]
## 리뷰 · 우선순위 64 / 80

draft임. `cursor/composer-2.5`가 툴 결과 이어가기에서 `resumeAction`으로 가면 빈 content + `finish_reason: stop`이거나 네이티브 툴만 다시 부름. `composer-2.5-fast`는 같은 경로가 됨.

픽스는 `cursorNeedsExternalToolContinuation()`으로 composer-2.5만 `userMessageAction` + 외부 툴 이어가기 텍스트로 보냄. `src/adapters/cursor/discovery.ts`랑 `protobuf-request.ts`. fast는 그대로.

모델 이름 특수 케이스가 늘면 깨지기 쉬움. 카탈로그 플래그로 빼는 게 더 안전함. 지금은 한 모델이라 참을 만함. 로컬 3/3은 있지만 CI에서 라이브 Cursor를 못 치면 회귀가 약함.

types.ts 스플릿 상관없음. #1742 이미지랑 다른 선임. 닫을 중복 아님.

**해결방안:** draft 해제 전에 composer-2.5 vs fast 분기 테스트 유지. 다른 composer 변형에 함부로 넓히지 말 것. 카탈로그 capability로 옮길지는 후속.

이 댓글은 grok-bot이 작성했습니다

**Review by Ingwannu — COMMENTED** [review]
Re-reviewed exact head `6648bd13cc4db89af4617eca6f65ade39506ba9a`.

The original source blocker is resolved: `cursorNeedsExternalToolContinuation()` is now confined to trailing-tool-result action selection, while ordinary `composer-2.5` root/history replay continues to use `isCursorExternalWireModel()`. The new ordinary-turn regression also exercises that boundary.

Local focused validation passed under the repository CPU limit:

- `bun test tests/cursor-blob.test.ts tests/cursor-discovery.test.ts tests/cursor-tool-continuation.test.ts tests/cursor-request-builder.test.ts`
- 116 passed, 0 failed

This is not merge-ready yet: the branch is still two commits behind current `dev`, the PR remains draft, and full exact-head CI has not run. Please rebase onto `03735eca62398c55056d4595145561aecc444e91`, complete the readiness checklist, and re-request review after the exact rebased head is green.

File: src/adapters/cursor/protobuf-request.ts [hunk 1]
```diff
@@ -793,8 +793,11 @@ function buildPreparedCursorRunRequest(
   const lastRawIsToolResult = request.rawMessages?.at(-1)?.role === "toolResult";
   // Native models resume the remembered Cursor conversation. External wire
   // models continue as userMessageAction so history-blob tool results stay
-  // visible without a ResumeAction.
-  const externalToolContinuation = lastRawIsToolResult && isCursorExternalWireModel(request.modelId);
+  // visible without a ResumeAction. Some native composer ids are also routed
+  // through the external continuation path (cursorNeedsExternalToolContinuation)
+  // because a bare resumeAction makes them continue exploring with native tools
+  // instead of answering (observed on composer-2.5; see discovery.ts).
+  const externalToolContinuation = lastRawIsToolResult && cursorNeedsExternalToolContinuation(request.modelId);
```
[the remark this answers is not in the document]
Comment (Ingwannu) [reply src/adapters/cursor/protobuf-request.ts:800]: I rechecked this against the exact current head. This finding should not be applied. composer-2.5 remains a Cursor-native model; the intended exception is limited to selecting userMessageAction for a trailing tool-result turn. Its native conversationTurns encoding still carries the native mcpToolCall plus result, while rootPromptMessagesJson already carries the model-visible tool-result text. Forcing the external-history branch would drop Composer-native tool state and would re-expand the exception beyond the action boundary that this PR intentionally narrowed. The focused native ordinary-turn and continuation regressions both pass, so no code change is requested here.

**Review by Ingwannu — APPROVED** [review]
Approved exact head b30e5baad3f60e16da84029e5e1da8fc9dd410be. The composer-2.5 exception is now correctly limited to trailing tool-result action selection; ordinary composer-2.5 and composer-2.5-fast retain their prior native replay semantics. I independently ran the four focused Cursor suites (116 passed, 0 failed) and repository typecheck successfully. I also reviewed the complete four-file fork diff and approved the pending Cross-platform CI and React Doctor runs. Do not merge until those exact-head workflows finish green.
